### PR TITLE
Switch audio framework from opensles to OHAudio and AVPlayer on openh…

### PR DESCRIPTION
**Pull request by zhaoyang57**
…armony platform.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -838,8 +838,8 @@ if(CC_USE_AUDIO)
         ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/OpenSLHelper.h
         ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/PcmAudioPlayer.cpp
         ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/PcmAudioPlayer.h
-        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/PcmAudioService.cpp
-        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/PcmAudioService.h
+        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/openharmony/PcmAudioService.cpp
+        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/openharmony/PcmAudioService.h
```
Comment (dumganhar) [inline CMakeLists.txt]: Why remove android/PcmAudioService.h/.cpp, will it break Android platform support?

Comment (dumganhar) [reply CMakeLists.txt]: It seems only openharmony platform use this CMakeLists.txt.  Right?

File: cocos/audio/openharmony/UrlAudioPlayer.cpp [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+        thiz->stop();
+        it++;
+    }
+}
+
+void UrlAudioPlayer::destroy()
+{
+    if (!*_isDestroyed)
+    {
+        *_isDestroyed = true;
```
Comment (minggo) [inline cocos/audio/openharmony/UrlAudioPlayer.cpp:352]: Though _isDestroyed is a shared ptr, but the usage here is not thread safe.

File: CMakeLists.txt (hunk 1 above)
Comment (zhaoyang57) [reply CMakeLists.txt]: right,  only openharmony platform use this CMakeLists.txt.

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 2 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp:352]: I think android should have the same issue. Do we need to add lock here?

File: cocos/audio/openharmony/UrlAudioPlayer.cpp [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+void UrlAudioPlayer::stopAll()
+{
+    __playerContainerMutex.lock();
+    auto temp = __playerContainer;
+    __playerContainerMutex.unlock();
+
+    std::map<OH_AVPlayer*, cocos2d::UrlAudioPlayer*>::iterator it = temp.begin();  
+    while(it != temp.end())  
+    {  
+        UrlAudioPlayer* thiz = (UrlAudioPlayer*)it->second;
```
Comment (qiuguohua) [inline cocos/audio/openharmony/UrlAudioPlayer.cpp]: The type of 'it ->sencond' is UrlAudioPlayer, so there is no need for type conversion here?

Comment (qiuguohua) [inline cocos/audio/openharmony/UrlAudioPlayer.cpp]: The type of 'it ->sencond' is UrlAudioPlayer, so there is no need for type conversion here?

File: cocos/audio/openharmony/UrlAudioPlayer.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    return _isLoop;
+}
+
+void UrlAudioPlayer::stopAll()
+{
+    __playerContainerMutex.lock();
+    auto temp = __playerContainer;
+    __playerContainerMutex.unlock();
+
+    std::map<OH_AVPlayer*, cocos2d::UrlAudioPlayer*>::iterator it = temp.begin();
```
Comment (qiuguohua) [inline cocos/audio/openharmony/UrlAudioPlayer.cpp]: No need to add a namespace here?

File: cocos/audio/openharmony/UrlAudioPlayer.cpp [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    __playerContainer[_playObj] = this;
+    __playerContainerMutex.unlock();
+
+    AVPlayerCallback callback;
+    callback.onInfo = this->onInfo;
+    OH_AVPlayer_SetPlayerCallback(_playObj, callback);
+    OH_AVPlayer_SetFDSource(_playObj, _assetFd->getFd(), start, length);
+    OH_AVErrCode code = OH_AVPlayer_Prepare(_playObj);
+    if (code == AV_ERR_OK) {
+        setState(State::INITIALIZED);
```
Comment (qiuguohua) [inline cocos/audio/openharmony/UrlAudioPlayer.cpp:306]: Do we need to set the default volume?

File: cocos/audio/openharmony/PcmAudioService.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+namespace cocos2d { 
+
+
+class AudioMixerController;
+
+class PcmAudioService
+{
+public:
+    inline int getChannelCount() const
```
Comment (qiuguohua) [inline cocos/audio/openharmony/PcmAudioService.h]: Can we put it on the same line?

File: cocos/audio/openharmony/PcmAudioService.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    virtual ~PcmAudioService();
+
+    bool init(AudioMixerController* controller, int numChannels, int sampleRate, int* bufferSizeInBytes);
+    
+
+    void pause();
+    void resume();
+
+    static int32_t AudioRendererOnWriteData(OH_AudioRenderer* renderer, void* userData, void* buffer, int32_t bufferLen);
+    int _numChannels;
```
Comment (qiuguohua) [inline cocos/audio/openharmony/PcmAudioService.h]: These variables are public?

File: cocos/audio/openharmony/PcmAudioService.cpp [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    ret = OH_AudioStreamBuilder_GenerateRenderer(_builder, &_audioRenderer);
+    if (ret != AUDIOSTREAM_SUCCESS) {
+        return false;
+    }
+
+    int32_t buffer_size;
+    OH_AudioRenderer_GetFrameSizeInCallback(_audioRenderer, &buffer_size);
+    _bufferSizeInBytes = buffer_size * 4;
```
Comment (qiuguohua) [inline cocos/audio/openharmony/PcmAudioService.cpp]: Why does this size need to be multiplied by 4?

File: cocos/audio/openharmony/PcmAudioService.cpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+
+int32_t PcmAudioService::AudioRendererOnWriteData(OH_AudioRenderer* renderer,
+    void* userData,
+    void* buffer,
+    int32_t bufferLen)
+{
+    auto *thiz = reinterpret_cast<PcmAudioService *>(userData);
+    if (bufferLen != thiz->_bufferSizeInBytes) {
+        __silenceData.resize(bufferLen, 0x00);
```
Comment (qiuguohua) [inline cocos/audio/openharmony/PcmAudioService.cpp:59]: Why do we need to resize the silent data, we shouldn't have changed it?

File: cocos/audio/openharmony/PcmAudioService.cpp [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            memcpy(buffer, current->buf, current->size < bufferLen ? current->size : bufferLen);
+        }
+    } else {
+        memcpy(buffer, __silenceData.data(), bufferLen);
+    }
+
+    return 0;
+}
+
+bool PcmAudioService::init(AudioMixerController *controller, int numChannels, int sampleRate, int *bufferSizeInBytes) {
```
Comment (qiuguohua) [inline cocos/audio/openharmony/PcmAudioService.cpp:81]: Can we try to keep the same parameters as android? Or do we add an extra parameter and add a comment?

File: cocos/audio/android/AudioMixerController.cpp [hunk 11]
```diff
@@ -50,6 +51,14 @@ AudioMixerController::AudioMixerController(int bufferSizeInFrames, int sampleRat
     _mixingBuffer.buf = memalign(32, _mixingBuffer.size);
     memset(_mixingBuffer.buf, 0, _mixingBuffer.size);
 }
+#elif CC_TARGET_PLATFORM == CC_PLATFORM_OPENHARMONY
+AudioMixerController::AudioMixerController(int sampleRate, int channelCount)
+: _sampleRate(sampleRate), _channelCount(channelCount), _mixer(nullptr), _isPaused(false), _isMixingFrame(false) {
+    ALOGV("In the constructor of AudioMixerController!");
+    int32_t maxBufferSize = 17832;
```
Comment (qiuguohua) [inline cocos/audio/android/AudioMixerController.cpp]: Can you use constants? Also can add comments why 17832?

File: cocos/audio/android/AudioEngine-inl.cpp [hunk 12]
```diff
@@ -199,7 +195,8 @@ bool AudioEngineImpl::init()
         // get the engine interface, which is needed in order to create other objects
         result = (*_engineObject)->GetInterface(_engineObject, SL_IID_ENGINE, &_engineEngine);
         if(SL_RESULT_SUCCESS != result){ ERRORLOG("get the engine interface fail"); break; }
-
+        
+    #if CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID
```
Comment (qiuguohua) [inline cocos/audio/android/AudioEngine-inl.cpp:199]: No need for _engineObject at all on the openharmony platform?

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 5 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp:306]: For avplayer, default volume is 1

File: cocos/audio/openharmony/PcmAudioService.cpp (hunk 8 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.cpp]: _bufferSizeInBytes is buffer length, and buffer_size is frameSize.
buffer length = frameSize * channel nums(2) *  bit depth (16bits, 2bytes)

File: cocos/audio/openharmony/PcmAudioService.cpp (hunk 10 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.cpp:81]: openharmony platform need add bufferSizeInBytes pointer param to get ohaudio bufferlength, which is needed by AudioMixer.

File: cocos/audio/android/AudioMixerController.cpp (hunk 11 above)
Comment (zhaoyang57) [reply cocos/audio/android/AudioMixerController.cpp]: For openharmony platform, buffer size is 4458 in normal latency mode, 240 in fast latency mode. so maxBufferLength = 4458 * 2(channel count) * 2(bit depth).

File: cocos/audio/android/AudioEngine-inl.cpp (hunk 12 above)
Comment (zhaoyang57) [reply cocos/audio/android/AudioEngine-inl.cpp:199]: It's needed by AudioDecoderSLES to decode audio files in some formats.
image

File: cocos/audio/openharmony/PcmAudioService.cpp (hunk 10 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.cpp:81]: comment added in PcmAudioService.h

File: cocos/audio/openharmony/PcmAudioService.cpp (hunk 9 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.cpp:59]: buffer size is 4458 in normal latency mode(bluetooth scence), 240 in fast latency mode(non bluetooth scence).

Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.cpp:59]: buffer size is 4458 in normal latency mode(bluetooth scence), 240 in fast latency mode(non bluetooth scence).

File: cocos/audio/openharmony/PcmAudioService.h (hunk 6 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.h]: done

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 3 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp]: fixed

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 4 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp]: fixed

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 3 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp]: fixed

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 4 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp:339]: fixed

File: cocos/audio/openharmony/PcmAudioService.h (hunk 7 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.h]: changed to private.

File: CMakeLists.txt [hunk 13]
```diff
@@ -848,8 +848,8 @@ if(CC_USE_AUDIO)
         ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/tinysndfile.h
         ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/Track.cpp
         ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/Track.h
-        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/UrlAudioPlayer.cpp
-        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/android/UrlAudioPlayer.h
+        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/openharmony/UrlAudioPlayer.cpp
+        ${CMAKE_CURRENT_LIST_DIR}/cocos/audio/openharmony/UrlAudioPlayer.h
```
Comment (qiuguohua) [inline CMakeLists.txt]: Can we put together files from the same platform?

File: cocos/audio/openharmony/UrlAudioPlayer.h [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    bool prepare(const std::string &url, std::shared_ptr<AssetFd> assetFd, int32_t start, int32_t length);
+
+    static void onInfo(OH_AVPlayer *player, AVPlayerOnInfoType type, int32_t extra);
+
+    static void stopAll();
+
+    void destroy();
+
+    inline void setState(State state)
+    { _state = state; };
```
Comment (qiuguohua) [inline cocos/audio/openharmony/UrlAudioPlayer.h]: On the same line?

File: cocos/audio/openharmony/UrlAudioPlayer.h [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+
+    virtual float getDuration() const override;
+
+    virtual float getPosition() const override;
+
+    virtual bool setPosition(float pos) override;
+
+    virtual void setPlayEventCallback(const PlayEventCallback &playEventCallback) override;
+
+    // Override Functions EndOv
```
Comment (qiuguohua) [inline cocos/audio/openharmony/UrlAudioPlayer.h]: Delete extra comments?

File: CMakeLists.txt (hunk 13 above)
Comment (zhaoyang57) [reply CMakeLists.txt]: done

File: cocos/audio/openharmony/UrlAudioPlayer.h (hunk 14 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.h]: done

File: cocos/audio/openharmony/UrlAudioPlayer.h (hunk 15 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.h]: done

**Review by qiuguohua — APPROVED** [review]

File: cocos/audio/openharmony/PcmAudioService.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    bool init(AudioMixerController* controller, int numChannels, int sampleRate, int* bufferSizeInBytes);
+    
+
+    void pause();
+    void resume();
+    
+private:
+    static int32_t AudioRendererOnWriteData(OH_AudioRenderer* renderer, void* userData, void* buffer, int32_t bufferLen);
+    int _numChannels;
+    int _sampleRate;
```
Comment (dumganhar) [inline cocos/audio/openharmony/PcmAudioService.h]: Where are `_numChannels`, `_sampleRate` used?

File: cocos/audio/openharmony/PcmAudioService.h [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    int _bufferSizeInBytes;
+
+    AudioMixerController* _controller;
+    OH_AudioRenderer *_audioRenderer;
+    OH_AudioStreamBuilder *_builder;
```
Comment (dumganhar) [inline cocos/audio/openharmony/PcmAudioService.h:62]: It's better to initialize member variables to a default value.

File: cocos/audio/openharmony/PcmAudioService.h (hunk 16 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.h]: Not used. Declaration removed.

File: cocos/audio/openharmony/PcmAudioService.h (hunk 17 above)
Comment (zhaoyang57) [reply cocos/audio/openharmony/PcmAudioService.h:62]: These pointers are dynamically assigned values in PcmAudioService.cpp

**Review by dumganhar — APPROVED** [review]

File: cocos/audio/openharmony/UrlAudioPlayer.cpp (hunk 2 above)
Comment (minggo) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp:352]: It depends on the usage:
- will this function be invoked in different threads?
- what will happen if `OH_AVPlayer_Reset(_playObj);` and `OH_AVPlayer_Release(_playObj);` be invoked more than one time?

Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp:352]: 1 It is possible to be called by different threads
2 When these interfaces are called repeatedly,  error code 3 will be returned, but no exception will be caused.

Comment (zhaoyang57) [reply cocos/audio/openharmony/UrlAudioPlayer.cpp:352]: image

**Review by minggo — APPROVED** [review]

**Pull request merged**